Eliminate need for root stacking context

mrobinson opened the pull request:
This makes the code quit a bit cleaner and allows for a greater variety
of display lists.

Fixes #1195.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/1211)
<!-- Reviewable:end -->

mrobinson commented:
cc @staktrace 

This eliminates the need for the root stacking context at the API level, although when one isn't provided we create one. This internal quirk can also be removed if it becomes a problem. I have chosen not to handle multiple root stacking contexts with blend isolation. We can make that work properly later, as well, if we decide that it's necessary.

mrobinson commented:
Also cc @Gankro, since these changes touch his new code a bit.

kvark commented:
Reviewed 18 of 18 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[webrender/src/frame.rs, line 666 at r1](https://reviewable.io:443/reviews/servo/webrender/1211#-KjdSvNBUIIQ_TdDdr8q:-KjdSvNCbnV0avIqtjjE:b43kyu3) ([raw file](https://github.com/servo/webrender/blob/9a22fffce0fd61c468d79f51b01404ade50780a6/webrender/src/frame.rs#L666)):*
> ```Rust
>                          reference_frame_relative_offset: LayerPoint) {
>         loop {
>             let subtraversal = {
> ```

that's still hacky but looks quite easier on the comprehension side
:+1: 

---

*[webrender/src/frame_builder.rs, line 259 at r1](https://reviewable.io:443/reviews/servo/webrender/1211#-KjdT8r5clHlYUx5IzrA:-KjdT8r5clHlYUx5IzrB:bmqy58k) ([raw file](https://github.com/servo/webrender/blob/9a22fffce0fd61c468d79f51b01404ade50780a6/webrender/src/frame_builder.rs#L259)):*
> ```Rust
>     }
> 
>     pub fn started_new_pipeline(&mut self) {
> ```

nit: I think this is a little against the naming recommendations? As in - I'd prefer the functions to start with a verb, as `start_new_pipeline`, or `notify_new_pipeline_started`

---

*[webrender_traits/src/api.rs, line 40 at r1](https://reviewable.io:443/reviews/servo/webrender/1211#-KjdViN0bWkxHtVTbwrg:-KjdViN0bWkxHtVTbwrh:bvlggz8) ([raw file](https://github.com/servo/webrender/blob/9a22fffce0fd61c468d79f51b01404ade50780a6/webrender_traits/src/api.rs#L40)):*
> ```Rust
>                    Epoch,
>                    PipelineId,
>                    LayoutSize, // viewport_size
> ```

we should consider making this enum variant (and some others) to have named fields. Not necessarily for this PR, but perhaps add a `TODO` in there?

---

*[webrender_traits/src/display_list.rs, line 440 at r1](https://reviewable.io:443/reviews/servo/webrender/1211#-KjdW3KRvKnMm8wE3vFV:-KjdW3KSlfTEidk6wiKj:b1pqj8p) ([raw file](https://github.com/servo/webrender/blob/9a22fffce0fd61c468d79f51b01404ade50780a6/webrender_traits/src/display_list.rs#L440)):*
> ```Rust
>         let start_time = precise_time_ns();
> 
>         // We start at 1 here, because the root scroll id is always 0.
> ```

would be nice to have this informal dependency formalized in code

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/1211)*
<!-- Sent from Reviewable.io -->

mrobinson commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[webrender/src/frame.rs, line 666 at r1](https://reviewable.io:443/reviews/servo/webrender/1211#-KjdSvNBUIIQ_TdDdr8q:-Kjg4khXftw4PpytwpOE:b-ucfknh) ([raw file](https://github.com/servo/webrender/blob/9a22fffce0fd61c468d79f51b01404ade50780a6/webrender/src/frame.rs#L666)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

that's still hacky but looks quite easier on the comprehension side
:+1: 
</blockquote></details>

Thanks!

---

*[webrender/src/frame_builder.rs, line 259 at r1](https://reviewable.io:443/reviews/servo/webrender/1211#-KjdT8r5clHlYUx5IzrA:-Kjg4Cioy8alvQCwszxQ:b6l8l1f) ([raw file](https://github.com/servo/webrender/blob/9a22fffce0fd61c468d79f51b01404ade50780a6/webrender/src/frame_builder.rs#L259)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

nit: I think this is a little against the naming recommendations? As in - I'd prefer the functions to start with a verb, as `start_new_pipeline`, or `notify_new_pipeline_started`
</blockquote></details>

You're right. I've changed this to `notify_waiting_for_root_stacking_context`.

---

*[webrender_traits/src/api.rs, line 40 at r1](https://reviewable.io:443/reviews/servo/webrender/1211#-KjdViN0bWkxHtVTbwrg:-Kjg4IRaKSZo2WBbscMo:b-etecvj) ([raw file](https://github.com/servo/webrender/blob/9a22fffce0fd61c468d79f51b01404ade50780a6/webrender_traits/src/api.rs#L40)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

we should consider making this enum variant (and some others) to have named fields. Not necessarily for this PR, but perhaps add a `TODO` in there?
</blockquote></details>

I've added a TODO here.

---

*[webrender_traits/src/display_list.rs, line 440 at r1](https://reviewable.io:443/reviews/servo/webrender/1211#-KjdW3KRvKnMm8wE3vFV:-Kjg4M4LtQoqLdV81Bgt:bkocfdu) ([raw file](https://github.com/servo/webrender/blob/9a22fffce0fd61c468d79f51b01404ade50780a6/webrender_traits/src/display_list.rs#L440)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

would be nice to have this informal dependency formalized in code
</blockquote></details>

I agree. Ideally it would be impossible to accidentally create an id that referred to the root scrolling node. I've been rolling the problem around in my head for a while and I think it's possible to do by modifying the definition of ClipId. In this case, I'm trying to make the issue clearer in the code by moving the magic value to a constant.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/1211)*
<!-- Sent from Reviewable.io -->

kvark commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/1211)*
<!-- Sent from Reviewable.io -->

kvark commented:
Thanks!
@bors-servo r+

bors-servo commented:
:pushpin: Commit 425155a has been approved by `kvark`

<!-- @bors-servo r=kvark 425155a2a5a7747b8f4119553c192ee17a520f45 -->

bors-servo commented:
:hourglass: Testing commit 425155a2a5a7747b8f4119553c192ee17a520f45 with merge 3035a6c1fd1e6f2110c38e919eaeab67c0669bb5...

bors-servo commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/servo/webrender/builds/230503456?utm_source=github_status&utm_medium=notification)
Approved by: kvark
Pushing 3035a6c1fd1e6f2110c38e919eaeab67c0669bb5 to master...

glennw commented:
@mrobinson I'm getting a number of panics in the Servo WPT tests when trying to do a WR update. The panic is unwrapping a stacking context index - in tests like ```meta_viewport_resize```. Is it likely related to this PR? It might just be that I ported the API changes incorrectly...?

mrobinson commented:
@glennw Thanks! I've uploaded https://github.com/servo/webrender/pull/1222 to fix this issue.

jrmuizel commented on webrender_traits/src/api.rs:
```diff
@@ -301,21 +303,23 @@ impl RenderApi {
+                            display_list_builder: DisplayListBuilder,
                             preserve_frame_state: bool) {
-        let (dl_data, dl_desc) = display_list.into_data();
+        let (pipeline_id, content_size, display_list) = display_list_builder.finalize();
```
This breaks Gecko. We finalize in a different process from the one that calls set_display_list. Is there a particular reason for this?

kvark replied:
#1226

staktrace commented:
@mrobinson FYI the change to set_display_list makes this hard to use with Gecko. We create a BuiltDisplayList which is serialized from one process to another across IPC and then we pass it to set_display_list. Your change to have that function take a DisplayListBuilder makes that harder, because now we have to wrap the BuiltDisplayList back into a DisplayListBuilder.

staktrace commented:
Whoops, didn't see jeff's comment just above. We're complaining about the same thing :)

staktrace commented:
Could the documentation for `set_display_list` also be updated to explain what `content_size` is supposed to be?

mrobinson commented:
@staktrace Created https://github.com/servo/webrender/pull/1243 for your request.

staktrace commented:
Thanks!
